### PR TITLE
[CARBONDATA-3310]fix IndexOutOfBoundException when drop table and create table run parallel for different table

### PR DESCRIPTION
**Pull request by BJangir**
### Problem:
when create table is fired and getTableFromMetadataCache is called to find the carbonTable from metadata, at the same time, if removeTableFromMetadata is called for any other table during table it removes one table from metadata and getTableFromMetadataCache throws IndexOutOFBoundException. 

### Solution:
maintain a boolean variable whether to remove from metadata during drop or not. If someone is finding table from metadata, do not remove at the same time.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed?
 NA
 - [x] Any backward compatibility impacted?
NA
 
 - [x] Document update required?
NA

 - [x] Testing done
tested with 100 times, with create and drop run parallely
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
NA

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2668/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10927/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2897/

**Comment by qiuchenjian** [comment]
[pr-3130](https://github.com/apache/carbondata/pull/3130)
Is it a same problem?

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala [hunk 1]
```diff
@@ -424,16 +436,21 @@ class CarbonFileMetastore extends CarbonMetaStore {
...
+        case Some(carbonTable) =>
+          metadata.carbonTables -= carbonTable
+        case None =>
+          if (LOGGER.isDebugEnabled) {
+            LOGGER.debug(s"No entry for table $tableName in database $dbName")
+          }
+      }
+      CarbonMetadata.getInstance.removeTable(dbName, tableName)
+    } else {
+      Thread.sleep(250)
```
Comment (qiuchenjian) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala]: Better to add timeout

**Comment by BJangir** [comment]
@qiuchenjian ,yes PR 3130 already has this fix So no need this PR.

**Pull request closed without merging**